decoder/expr_function.go

```diff
@@ -55,6 +56,61 @@ func (fe functionExpr) CompletionAtPos(ctx context.Context, pos hcl.Pos) []lang.
 
 		prefix := rootName[0:prefixLen]
 		return fe.matchingFunctions(prefix, eType.Range())
+
+	case *hclsyntax.ExprSyntaxError:
+		// TODO: "provider:" does not jump here, seems to be some other expression?
+
+		// This range can range up until the end of the file in case of invalid config
+		if eType.SrcRange.ContainsPos(pos) {
+			// we are somewhere in the range for this attribute but we don't have an expression range to check
+			// so we look back to check whether we are in a partially written provider defined function
+			fileBytes := fe.pathCtx.Files[eType.SrcRange.Filename].Bytes
+
+			recoveredPrefixBytes := recoverLeftBytes(fileBytes, pos, func(offset int, r rune) bool {
+				return !isNamespacedFunctionNameRune(r)
+			})
+			// recoveredPrefixBytes also contains the rune before the function name, so we need to trim it
+			_, lengthFirstRune := utf8.DecodeRune(recoveredPrefixBytes)
+			recoveredPrefixBytes = recoveredPrefixBytes[lengthFirstRune:]
+
+			recoveredSuffixBytes := recoverRightBytes(fileBytes, pos, func(offset int, r rune) bool {
+				return !isNamespacedFunctionNameRune(r) && r != '('
+			})
+			// recoveredSuffixBytes also contains the rune after the function name, so we need to trim it
+			_, lengthLastRune := utf8.DecodeLastRune(recoveredSuffixBytes)
+			recoveredSuffixBytes = recoveredSuffixBytes[:len(recoveredSuffixBytes)-lengthLastRune]
+
+			recoveredIdentifier := append(recoveredPrefixBytes, recoveredSuffixBytes...)
+
+			// TODO: this is specific to Terraform provider defined functions, we should generalize this
```

Comment:
We are currently only looking for the `provider:` prefix, but in theory `hcl-lang` should support any namespaced function, right?

Reply:
Yes, we should avoid implementing anything Terraform specific in `hcl-lang`. Any namespaced function should work and also the short syntax `namespace::func()`